### PR TITLE
Add deepsearch implementation

### LINKED ISSUE #7: Add Deepsearch implementation

### LINKED ISSUE #29: Add deepsearch (Tiefensuche) implementation

### PR DESCRIPTION
**Pull request by cbl**
This pr adds the implementation for the deepsearch algorithm.

To display a more readable result for the `discovered` and `finished` times of the nodes, a `TablePrinter` printer was added as well.

## Example

The deepsearch algorithm data requires an instance of a graph and an array containing the node names:

```java
int size = 4;
String[] nodeNames = {"A", "B", "C", "D"};
Graph<Integer, boolean[][]> graph = new AdjacentMatrixGraph(size);

graph.setEdge(0, 3);
graph.setEdge(0, 2);
graph.setEdge(1, 3);
graph.setEdge(2, 1);
graph.setEdge(3, 2);

final Algorithm<Event, Deepsearch.Data> a = new Deepsearch();
final EventConsumer<Event> ec = new GeneralEventConsumer();

a.run(ec, new Deepsearch.Data(graph, nodeNames));

ec.visitEvents(new LogEventVisitor());
```

The example will resullt in the following output:

Bildschirmfoto 2021-06-13 um 20 08 54

closes https://github.com/cbl/algorithm-analyzer/issues/7

### CODE REVIEW

**Comment by cbl** [comment]
@herrmanno 

What do you mean by this?

https://github.com/cbl/algorithm-analyzer/blob/d72ba6ecbf2d8f037783d7a0652bad0a93170059/src/main/java/com/github/cbl/algorithm_analyzer/graphs/AdjacentMatrixGraph.java#L8

**Comment by herrmanno** [comment]
> @herrmanno
> 
> What do you mean by this?
> 
> https://github.com/cbl/algorithm-analyzer/blob/d72ba6ecbf2d8f037783d7a0652bad0a93170059/src/main/java/com/github/cbl/algorithm_analyzer/graphs/AdjacentMatrixGraph.java#L8

Thought it would be a good idea to implement `Graph` as `Graph<T>`, being polymorphic over the type of its nodes.
But, especially when implemented as array-based adjacent matrix, this may be a disappropriate overhead.

File: src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    protected EdgeType getEdgeType(
+            int from,
+            int to,
+            Graph<Integer, boolean[][]> graph,
+            int[] discovered,
+            int[] finished,
+            int[] previous) {
+        if (!graph.hasEdge(from, to)) {
```
Comment (herrmanno) [inline src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java:100]: Just a minor note:
When handling multiple mutual exclusive cases, why not wrap it all together in a form of

```java
if (…) {
…
} else if (…) {
…
} else if (…) {
…
} else {
…
}
```

Eases the reading flow a little bit, in my opinion.
But of course not a real problem here, at worst a very slight code smell (but again: talking about opinions, not facts).

File: src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import com.github.cbl.algorithm_analyzer.util.TablePrinter;
+
+import java.util.StringJoiner;
+
+public class Deepsearch implements Algorithm<Event, Deepsearch.Data> {
+
+    public static record Data(Graph<Integer, boolean[][]> graph, String[] nodeNames) {}
+    ;
+
+    public static record FinalTimesEvent(Data data, int[] discovered, int[] finished)
```
Comment (herrmanno) [inline src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java]: Maybe also add a row of 'previous node'? Could not hurt.

**Review by herrmanno — APPROVED** [review]
Looks good. I haven't checked the core of the algorithm but I'm sure, it works if expected.
Maybe just add one or two unit tests based on past submissions of ours.

**Comment by cbl** [comment]
I improved the look of the table and added `Vorgaenger` to the list.

Bildschirmfoto 2021-06-13 um 22 08 29

I am relatively confident that the implementation is correct, as Prof. Dr. Weicker gave me 30 points for the task

File: src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java (hunk 1 above)
Comment (cbl) [reply src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java:100]: Personally, I prefer early returns to if-else conditions, although I haven't been consistent here either.

Comment (cbl) [reply src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java:100]: I added early returns for all conditions for consistency https://github.com/cbl/algorithm-analyzer/commit/ac6862303d12ff75565f0eed68858cd4e0e67631

**Pull request merged**